### PR TITLE
Refactor root cap

### LINKED ISSUE #197: [Feature Request]: DAOSpace Cap refactor
# Feature Request

**Describe the Feature Request**

**Describe Preferred Solution**

**Describe Alternatives**

**Related Code**

**Additional Context**

**If the feature request is approved, would you be willing to submit a PR?**
Yes / No _(Help can be provided if you need assistance submitting a PR)_

### PR DESCRIPTION
**Pull request by 0xpause**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: resolve #197

## What is the new behavior?

- Remove DAORootCap and the *_with_root_cap functions. 
- DAOs can use the struct type `DAOT` as a special plugin type to get all other caps. In this case, no need to set `required_caps` for this special "plugin".

### CODE REVIEW

File: sources/daospace/DAOSpace.move [hunk 1]
```diff
@@ -272,7 +269,15 @@ module StarcoinFramework::DAOSpace {
...
 
     /// Create a dao with a exists DAO account
-    public fun create_dao<DAOT: store>(cap: DAOAccountCap, name: vector<u8>, image_data:Option::Option<vector<u8>>, image_url:Option::Option<vector<u8>>, description: vector<u8>, ext: DAOT, config: DAOConfig): DAORootCap<DAOT> acquires DAOEvent {
+    public fun create_dao<DAOT: store>(
+        cap: DAOAccountCap,
+        name: vector<u8>,
+        image_data:Option::Option<vector<u8>>,
+        image_url:Option::Option<vector<u8>>,
+        description: vector<u8>,
+        ext: DAOT,
```
Comment (yubing744) [inline sources/daospace/DAOSpace.move:278]: 感觉 DAOT作为 cap 和 _witness，就不太适合作为DAO的扩展字段了，不然每次构造cap或者_witness都需要补全DAOT中定义的字段

File: sources/daospaceplugin/StakeToSBTPlugin.move [hunk 2]
```diff
@@ -116,7 +116,7 @@ module StarcoinFramework::StakeToSBTPlugin {
     }
 
     /// Accept token with token type by given DAO root capability
```
Comment (yubing744) [inline sources/daospaceplugin/StakeToSBTPlugin.move:118]: 注释没改到，还是DAO root capability

File: sources/daospaceplugin/StakeToSBTPlugin.move [hunk 3]
```diff
@@ -133,8 +133,8 @@ module StarcoinFramework::StakeToSBTPlugin {
     }
 
     /// Set sbt weight by given DAO root capability
```
Comment (yubing744) [inline sources/daospaceplugin/StakeToSBTPlugin.move:135]: 这里也是：注释没改到，还是DAO root capability

File: sources/daospace/DAOSpace.move [hunk 4]
```diff
@@ -841,21 +836,7 @@ module StarcoinFramework::DAOSpace {
...
-            dao_id: dao_id(dao_address),
-            type:MEMBERJOIN_DIRECT,
-            member_id:Option::destroy_some(query_member_id<DAOT>(to_address)),
-            addr:to_address,
-            sbt: init_sbt,
-        });
-    }
-
-    // TODO refactor the cap and remove this function
+    // join member
```
Comment (jolestar) [inline sources/daospace/DAOSpace.move:839]: 这里的 `to_address: address` 是不是应该变成 `signer`？这样可以把接受 NFT 的逻辑也加进来。这个方法的场景是有用户的 signer 同时有 cap，否则就应该通过 offer 的方式。

File: sources/daospace/DAOSpace.move (hunk 1 above)
Comment (jolestar) [reply sources/daospace/DAOSpace.move:278]: 也有道理，这个 ext 好像也没想好使用场景，要不就干掉吧。如果开发者一定要存数据，可以通过 DAO 提供的 storage 接口

File: sources/daospaceplugin/StakeToSBTPlugin.move (hunk 3 above)
Comment (0xpause) [reply sources/daospaceplugin/StakeToSBTPlugin.move]: fixed

File: sources/daospace/DAOSpace.move (hunk 1 above)
Comment (0xpause) [reply sources/daospace/DAOSpace.move]: 已修改

File: sources/daospace/DAOSpace.move (hunk 4 above)
Comment (0xpause) [reply sources/daospace/DAOSpace.move:813]: 已修改

**Review by jolestar — APPROVED** [review]

**Pull request merged**